File: 07-Encordados-JPA-Hibernate-Java/src/main/java/models/Tenista.java
package models;

import enums.ManoDominante;
import enums.ManoDominanteUtils;
import enums.TipoReves;
import enums.TipoRevesUtils;
import jakarta.persistence.*;
import org.hibernate.annotations.UuidGenerator;
import utils.MyLocale;

import java.time.LocalDate;
import java.util.Objects;
import java.util.UUID;


@Entity
@Table(name = "tenistas") // Solo es necexario si cambia la tabla respecto al modelo
@NamedQuery(name = "Tenista.findAll", query = "SELECT t FROM Tenista t")
@NamedQuery(name = "Tenista.porMarcaRaqueta", query = "SELECT t FROM Tenista t WHERE t.raqueta.marca = ?1")

public class Tenista {
    private UUID uuid = UUID.randomUUID();
    private String nombre;
    private Integer ranking;
    private LocalDate fechaNacimiento;
    private Integer añoProfesional;
    private Integer altura;
    private Integer peso;
    private Double ganancias;
    private ManoDominante manoDominante;
    private TipoReves tipoReves;
    private Integer puntos;
    private String pais;
    // Mi referencia a la raqueta
    private Raqueta raqueta;

    public Tenista(String nombre, Integer ranking, LocalDate fechaNacimiento, Integer añoProfesional, Integer altura,
                   Integer peso, Double ganancias, ManoDominante manoDominante, TipoReves tipoReves,
                   Integer puntos, String pais) {
        this.nombre = nombre;
        this.ranking = ranking;
        this.fechaNacimiento = fechaNacimiento;
        this.añoProfesional = añoProfesional;
        this.altura = altura;
        this.peso = peso;
        this.ganancias = ganancias;
        this.manoDominante = manoDominante;
        this.tipoReves = tipoReves;
        this.puntos = puntos;
        this.pais = pais;
    }

    public Tenista() {
    }

    //
//    @Id
//    @GeneratedValue(strategy = GenerationType.IDENTITY)
//    @Column(name = "id")
    @Id
    @GeneratedValue
    @UuidGenerator // Genera un UUID
    @Column(name = "uuid")
    public UUID getUuid() {
        return uuid;
    }

    private void setUuid(UUID uuid) {
        this.uuid = uuid;
    }


    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public Integer getRanking() {
        return ranking;
    }

    public void setRanking(Integer ranking) {
        this.ranking = ranking;
    }

    public LocalDate getFechaNacimiento() {
        return fechaNacimiento;
    }

    public void setFechaNacimiento(LocalDate fechaNacimiento) {
        this.fechaNacimiento = fechaNacimiento;
    }

    public Integer getAñoProfesional() {
        return añoProfesional;
    }

    public void setAñoProfesional(Integer añoProfesional) {
        this.añoProfesional = añoProfesional;
    }

    public Integer getAltura() {
        return altura;
    }

    public void setAltura(Integer altura) {
        this.altura = altura;
    }

    public Integer getPeso() {
        return peso;
    }

    public void setPeso(Integer peso) {
        this.peso = peso;
    }

    public Double getGanancias() {
        return ganancias;
    }

    public void setGanancias(Double ganancias) {
        this.ganancias = ganancias;
    }

    public ManoDominante getManoDominante() {
        return manoDominante;
    }

    public void setManoDominante(ManoDominante manoDominante) {
        this.manoDominante = manoDominante;
    }

    public TipoReves getTipoReves() {
        return tipoReves;
    }

    public void setTipoReves(TipoReves tipoReves) {
        this.tipoReves = tipoReves;
    }

    public Integer getPuntos() {
        return puntos;
    }

    public void setPuntos(Integer puntos) {
        this.puntos = puntos;
    }

    public String getPais() {
        return pais;
    }

    public void setPais(String pais) {
        this.pais = pais;
    }


    // Relación 1-N con Raqueta, un tenista puede tener una raqueta y una raqueta puede tener muchos tenistas
    @ManyToOne
    @JoinColumn(name = "raqueta_uuid", referencedColumnName = "uuid", nullable = false)
    public Raqueta getRaqueta() {
        return raqueta;
    }

    public void setRaqueta(Raqueta raqueta) {
        this.raqueta = raqueta;
    }


    public void removeRaqueta() {
        this.raqueta.getTenistas().remove(this);
        this.raqueta = null;
    }

    @Override
    public String toString() {
        return "Tenista{" +
                "uuid=" + uuid +
                ", nombre='" + nombre + '\'' +
                ", ranking=" + ranking +
                ", fechaNacimiento=" + MyLocale.toLocalDate(fechaNacimiento) +
                ", añoProfesional=" + añoProfesional +
                ", altura=" + altura +
                ", peso=" + peso +
                ", ganancias=" + MyLocale.toLocalMoney(ganancias) +
                ", manoDominante=" + ManoDominanteUtils.toString(manoDominante) +
                ", tipoReves=" + TipoRevesUtils.toString(tipoReves) +
                ", puntos=" + puntos +
                ", pais='" + pais + '\'' +
                ", raqueta=" + raqueta +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Tenista)) return false;
        Tenista tenista = (Tenista) o;
        return uuid.equals(tenista.uuid) && nombre.equals(tenista.nombre) && ranking.equals(tenista.ranking) && fechaNacimiento.equals(tenista.fechaNacimiento) && añoProfesional.equals(tenista.añoProfesional) && altura.equals(tenista.altura) && peso.equals(tenista.peso) && ganancias.equals(tenista.ganancias) && manoDominante == tenista.manoDominante && tipoReves == tenista.tipoReves && puntos.equals(tenista.puntos) && pais.equals(tenista.pais);
    }

    @Override
    public int hashCode() {
        return Objects.hash(uuid, nombre, ranking, fechaNacimiento, añoProfesional, altura, peso, ganancias, manoDominante, tipoReves, puntos, pais);
    }
}

File: 08-Encordados-JPA-Hibernate-Kotlin/src/main/kotlin/models/Tenista.kt
package models

import extensions.toLocalDate
import extensions.toLocalMoney
import extensions.toLocalNumber
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.GenericGenerator
import org.hibernate.annotations.Type
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "tenistas") // Solo es necesario si cambia la tabla respecto al modelo
// Defino estas queries que estarán disponibles globalmente
// @NamedQuery(name = "Tenista.findAll", query = "SELECT t FROM Tenista t")
// @NamedQuery(name = "Tenista.porMarcaRaqueta", query = "SELECT t FROM Tenista t WHERE t.raqueta.marca = ?1")
@NamedQueries(
    NamedQuery(name = "Tenista.findAll", query = "SELECT t FROM Tenista t"),
    NamedQuery(
        name = "Tenista.porMarcaRaqueta",
        query = "SELECT t FROM Tenista t WHERE t.raqueta.marca = ?1"
    ),
)
data class Tenista(
    @Id @GeneratedValue //@UuidGenerator // Hibernate 6
    // Hibernate 5
    @GenericGenerator(
        name = "UUID",
        strategy = "org.hibernate.id.UUIDGenerator",
    )
    @Column(name = "uuid")
    @Type(type = "uuid-char")
    val uuid: UUID = UUID.randomUUID(),
    var nombre: String,
    var ranking: Int,
    var fechaNacimiento: LocalDate,
    var añoProfesional: Int,
    var altura: Int,
    var peso: Int,
    var ganancias: Double,
    var manoDominante: ManoDominante,
    var tipoReves: TipoReves,
    var puntos: Int,
    var pais: String,
    // Relación 1-N con Raqueta, un tenista puede tener una raqueta y una raqueta puede tener muchos tenistas
    @ManyToOne
    @JoinColumn(name = "raqueta_uuid", referencedColumnName = "uuid", nullable = true)
    var raqueta: Raqueta? = null, // No tiene por que tener raqueta

    @Column(name = "created_at")
    @Type(type = "org.hibernate.type.LocalDateTimeType")
    @CreationTimestamp
    val createdAt: LocalDateTime = LocalDateTime.now(),
) {

    override fun toString(): String {
        return "Tenista(uuid=$uuid, nombre='$nombre', ranking=$ranking, " +
                "fechaNacimiento=${fechaNacimiento.toLocalDate()}, " +
                "añoProfesional=$añoProfesional, " +
                "altura=${(altura.toDouble() / 100).toLocalNumber()} cm, " +
                "peso=$peso Kg, " +
                "ganancias=${ganancias.toLocalMoney()}, " +
                "manoDominante=${manoDominante.mano}, " +
                "tipoReves=${tipoReves.tipo}, " +
                "puntos=$puntos, pais=$pais, " +
                // Cuidado con la recursividad, porque al imprimir la raqueta, esta a su vez imprime el tenista y
                // así sucesivamente entramos en un bucle infinito
                "raqueta=${raqueta?.uuid})"
    }

    // ENUMS de la propia clase
    enum class ManoDominante(val mano: String) {
        DERECHA("DERECHA"),
        IZQUIERDA("IZQUIERDA");

        companion object {
            fun from(manoDominante: String): ManoDominante {
                return when (manoDominante.uppercase()) {
                    "DERECHA" -> DERECHA
                    "IZQUIERDA" -> IZQUIERDA
                    else -> throw IllegalArgumentException("ManoDominante no reconocida")
                }
            }
        }
    }

    enum class TipoReves(val tipo: String) {
        UNA_MANO("UNA MANO"),
        DOS_MANOS("DOS MANOS");

        companion object {
            fun from(tipoReves: String): TipoReves {
                return when (tipoReves.uppercase()) {
                    "UNA MANO" -> UNA_MANO
                    "DOS MANOS" -> DOS_MANOS
                    else -> throw IllegalArgumentException("TipoReves no reconocida")
                }
            }
        }
    }


}